### PR TITLE
Fix support for multi node JAX sharding

### PR DESCRIPTION
**Pull request by awolant**
## Category:

**Bug fix**

## Description:

In some situations `data_iterator` for JAX did not work well in multiprocess environment. This PR improves that. 

### Affected modules and functionalities:

Iterator for JAX. Some adjustments were mode on a code path where `sharding` argument is provided. 

### Tests:

- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3670

### CODE REVIEW

File: dali/python/nvidia/dali/plugin/jax/iterator.py [hunk 1]
```diff
@@ -172,7 +172,7 @@ def _next_impl(self):
         for category_id, category_name in enumerate(self.output_map):
             category_outputs = self._gather_outputs_for_category(pipelines_outputs, category_id)
 
-            if self._num_gpus == 1:
+            if self._num_gpus == 1 and self._sharding is None:
```
Comment (awolant) [inline dali/python/nvidia/dali/plugin/jax/iterator.py:175]: `self._num_gpus` is equal to the number of pipelines run by this instance of the iterator. So if we need to distinguish multi node training with one GPU per node (process) from just one GPU training.

File: dali/python/nvidia/dali/plugin/jax/iterator.py [hunk 2]
```diff
@@ -229,7 +229,12 @@ def _build_output_with_sharding(self, category_outputs):
         This output is compatible with automatic parallelization with JAX.
         """
         shard_shape = category_outputs[0].shape
-        global_shape = (self._num_gpus * shard_shape[0], *shard_shape[1:])
+
+        if isinstance(self._sharding, NamedSharding):
+            global_shape = (self._sharding.mesh.size * shard_shape[0], *shard_shape[1:])
+        else:
+            global_shape = (self._sharding.shape[0] * shard_shape[0], *shard_shape[1:])
```
Comment (awolant) [inline dali/python/nvidia/dali/plugin/jax/iterator.py:236]: `sharding` variants have inconsisten APIs when it comes to getting the global shape.

**Comment by awolant** [comment]
!build

**Review by banasraf — APPROVED** [review]

File: dali/python/nvidia/dali/plugin/jax/iterator.py [hunk 3]
```diff
@@ -303,9 +308,17 @@ def create_iterator(*args, **wrapper_kwargs):
...
 
-                devices_to_use = devices if devices is not None else jax.devices()
+                devices_to_use = devices if devices is not None else jax.local_devices()
+                num_shards = len(devices) if devices is not None else jax.device_count()
+
+                if devices is not None:
+                    assert jax.local_device_count() == jax.device_count(), (
+                        "Iterator compatible with pmapped JAX functions does not support "
+                        "running in multiprocess mode. Use `sharding` argument instead."
+                    )
```
Comment (klecki) [inline dali/python/nvidia/dali/plugin/jax/iterator.py:321]: Assertions != error checking. If you intend this to be a proper error, please use appropriate exception with explicit raise.

Comment (awolant) [reply dali/python/nvidia/dali/plugin/jax/iterator.py]: Done

**Review by klecki — APPROVED** [review]

**Comment by awolant** [comment]
!build

**Comment by awolant** [comment]
!build

**Comment by awolant** [comment]
!build

**Comment by awolant** [comment]
!build

**Pull request merged**